torchvision/transforms/v2/functional/_utils.py

```diff
@@ -6,3 +7,25 @@
 
 def is_simple_tensor(inpt: Any) -> bool:
     return isinstance(inpt, torch.Tensor) and not isinstance(inpt, Datapoint)
+
+
+_KERNEL_REGISTRY = {}
+
+
+def register_kernel(dispatcher, datapoint_cls, *, datapoint_wrapping=True):
+    def datapoint_wrapper(kernel):
+        @functools.wraps(kernel)
+        def wrapper(inpt, *args, **kwargs):
+            return type(inpt).wrap_like(inpt, kernel(inpt.as_subclass(torch.Tensor), *args, **kwargs))
+
+        return wrapper
```

Comment:
Right now, this is just a helper to avoid the boiler-plate unwrapping and re-wrapping. However, we could expand this in two ways to be able to use the default behavior in 99% of the cases, without implementing an intermediate layer:

1. We could a `select_parameters` parameter to `register_kernel` that covers the use case if the kernel we want to register only uses a subset of parameters of the dispatcher.
2. We could add automatic handling for datapoints that carry metadata (currently only `BoundingBox`):
   1. We can inspect the `kernel` to find metadata that needs to be passed explicitly. This can happen when registering and thus only once and without runtime performance regression. At runtime we use this information and adjust the `kwargs` before calling the kernel.
   2. We can handle returned metadata. For `BoundingBox` we know that the kernel output is either `Tensor` or `Tuple[Tensor, Tuple[int, int]]`. In the second case, the second element in the returned tuple is the `spatial_size`, which we could in the returned `BoundingBox`.

Going for that would make our implementation quite a bit simpler. However this also has two downsides:

1. Users that look at our code on how to write custom kernels for their datapoints, will likely miss that their is an intermediate layer in the first place.
2. The automatic returned metadata handling (2.ii above) will only work for builtin datapoints.

Thoughts?

Reply:
The datapoint_warpping logic (and the rest) should be kept outside of the registration mechanism IMO. I understand it makes it convenient to have the wrapping here, but these are 2 very distinct concepts.

The wrapping logic should probably just be a different decorator.


Also... why do we even wrap again? Was it for perf reasons? Is it *that* much better?

Reply:
> Also... remind me why we need to wrap/unwrap? Was it for perf reasons? Is it that much better?

There two parts here:

1. We unwrap before we call into a kernel for performance reasons. Whenever perform any operation on a datapoint, `__torch_function__` is invoked. This will even happen multiple times if the the operation is one that retains the type, like `.clone()`. IIRC, we are talking about low double digit microseconds here. However, this is not needed for functionality.
2. We wrap the output of the kernel to retain the input type of the dispatcher. Basically if you put in a `BoundingBox` you also want to have one back rather than a simple tensor. Otherwise, we force the user to always wrap by themselves in between calls.

> The datapoint_warpping logic (and the rest) should be kept outside of the registration mechanism IMO. I understand it makes it convenient to have the wrapping here, but these are 2 very distinct concepts.
> 
> The wrapping logic should probably just be a different decorator.

Although I agree that these are two concepts and should be kept separate, I don't think we can. The wrapping must only happen for when the kernel is called from a dispatcher, but not when called directly. Meaning

```python
@register_kernel(resize, datapoints.Image)
@default_datapoint_wrapping
def resize_image_tensor(...)
    ...
```

is wrong, since now calling `resize_image_tensor` will also try to wrap although it shouldn't. It will also fail if we pass a plain tensor, since the re-wrapping on the way back requires the input to be a datapoint.

The only way to go with two decorators is to never actually register a kernel, but always an intermediate function:

```python
@register_kernel(resize, datapoints.Image)
@default_datapoint_wrapping
def _resize_image_tensor_dispatch(...)
    return resize_image_tensor(...)
```

That would basically mean that we just move exactly what we have on the methods now as functions into functional modules. IMHO this is too much boilerplate to justify keeping both functionalities separate. Plus, we also need to find a new name for `register_kernels` in that case, because what we register doesn't really fit our definition or kernel anymore.